src/sdam/state/mod.rs

```diff
@@ -2,161 +2,302 @@ pub(super) mod server;
+#[derive(Clone, Debug)]
+pub(crate) struct Topology {
+    state: Arc<RwLock<TopologyState>>,
+    common: Common,
 }
 
-impl TopologyUpdateCondvar {
-    pub(crate) fn new() -> Self {
-        Self {
-            condvar: Arc::new(Condvar::new()),
-            mutex: Default::default(),
-        }
-    }
-
-    pub(crate) fn wait_timeout(&self, duration: Duration) {
-        let _ = self
-            .condvar
-            .wait_timeout(self.mutex.lock().unwrap(), duration)
-            .unwrap();
-    }
-
-    fn notify(&self) {
-        self.condvar.notify_all()
-    }
+/// A weak reference to the topology, which includes the current state as well as the client
+/// options and the message manager.
+#[derive(Clone, Debug)]
+pub(crate) struct WeakTopology {
+    state: Weak<RwLock<TopologyState>>,
+    common: Common,
 }
 
-/// Contains the SDAM state for a Client.
-#[derive(Clone, Derivative)]
-#[derivative(Debug)]
-pub(crate) struct Topology {
-    /// The SDAM and server selection state machine.
-    pub(crate) description: TopologyDescription,
-
-    /// The state associated with each server in the cluster.
-    pub(crate) servers: HashMap<StreamAddress, Arc<Server>>,
-
-    #[derivative(Debug = "ignore")]
-    condvar: TopologyUpdateCondvar,
-
+/// Encapsulates the common elements of Topology and WeakTopology, which includes the message
+/// manager and the client options.
+#[derive(Clone, Debug)]
+struct Common {
+    message_manager: TopologyMessageManager,
     options: ClientOptions,
 }
 
+/// The current state of the topology, which includes the topology description and the set of
+/// servers.
+#[derive(Clone, Debug)]
+pub(crate) struct TopologyState {
+    description: TopologyDescription,
+    servers: HashMap<StreamAddress, Arc<Server>>,
+}
+
 impl Topology {
-    /// Creates a new Topology given the `options`. Arc<RwLock<Topology> is returned rather than
-    /// just Topology so that monitoring threads can hold a Weak reference to it.
-    pub(crate) fn new(
-        condvar: TopologyUpdateCondvar,
-        mut options: ClientOptions,
-    ) -> Result<Arc<RwLock<Self>>> {
+    /// Creates a new Topology given the `options`.
+    pub(crate) async fn new(mut options: ClientOptions) -> Result<Self> {
         let description = TopologyDescription::new(options.clone())?;
+
         let hosts: Vec<_> = options.hosts.drain(..).collect();
 
-        let topology = Arc::new(RwLock::new(Topology {
+        let common = Common {
+            message_manager: TopologyMessageManager::new(),
+            options: options.clone(),
+        };
+
+        let state = Arc::new(RwLock::new(TopologyState {
             description,
             servers: Default::default(),
-            condvar,
-            options,
         }));
 
+        let topology = Topology { common, state };
+
         {
-            let mut topology_lock = topology.write().unwrap();
+            let mut state = topology.state.write().await;
```

Comment:
Now that this is happening on the separate state struct, we can do this initialization before putting it behind the lock.

Reply:
Ah, good point!

Reply:
Actually, we can't do this easily, since we need to have the already created `Topology` to pass to `add_new_server` to be able to start the monitoring threads. Moving the lock to a separate struct doesn't really change anything.

Reply:
Oh right, forgot about that.